### ISSUE TITLE
Implement SignCLIP Distances metric

### ISSUE THREAD

**Issue by cleong110** [sign-language-processing/pose-evaluation#1]
Implement a metric for SignCLIP embedding distances. 

As in, take two poses and a SignCLIP model, embed them both, calculate cosine similarity.

[Reference implementation with regular CLIP](https://github.com/sign-language-processing/signwriting-evaluation/blob/main/signwriting_evaluation/metrics/clip.py)

TODO:
- [ ] A separate Eval script that loads in saved embeddings
- [ ] Read/retrieve embeddings to a diskcache
- [ ] Support for different devices, cpu and cuda, (unless this should be higher in the hierarchy? DistanceMetric?)
- [ ] Support quick batch cosine distances via torch

**Comment by cleong110** [comment]
I previously implemented this by saving embeddings to a pgvector database and then doing an Order By. https://github.com/pgvector/pgvector-python?tab=readme-ov-file#peewee

**Comment by cleong110** [comment]
There are two major challenges to address for using my implementation for a metric.
1. generating the SignCLIP embeddings. Right now the process is quite manual and requires me to download the SignCLIP repo and modify some code so I can run it. Once that's done I can save off the resulting vectors as .npy files.
2. the reliance on pgvector database. It's cumbersome to have to set up a whole postgresql database every time.

**Comment by cleong110** [comment]
The second one is solvable by simply loading in the .npy files from step 1 directly, then calculating cosine distances. 

The first one I'm not sure what to do about other than updating the setup scripts and README files

**Comment by cleong110** [comment]
https://github.com/cleong110/pose-evaluation/tree/signclip_metric has a basic implementation, based on loading in .npy files. It can calculate about 5k scores per second. 

I have 83116 .npy files per signclip model saved off. math.comb(83116, 2) shows that there are about 3.5 billion combinations, I might want to fix a few things. Most notably: 
* don't save the full ID for hyp, ref in a .csv, every letter counts.
* Maybe use ProcessPoolExecutor to parallelize? Make better use of all my cpus.

**Comment by AmitMY** [comment]
I looked at the implementation. I think you should use a cache, similar to what I have for the CLIP metric. it will be faster to retrieve the embedding from memory than from disk every single time

also, i would argue that you should not implement `score` but instead implement `score_all` (again, same as my implementation), that way, you can parallelize all of the scoring by doing large matrix multiplication (and it can happen on GPU so it will be so much faster)

**Comment by AmitMY** [comment]
## Quick calculation: 
SignCLIP embeddings are `D=768` dimensions
to calculate cosine similarity you need D multiplications + norm multiplication + one division 

A modern GPU, like RTX 4090 supports (at 100% efficiency) 82.58 TFLOPS which are 82,580,000,000,000 FLOPS, meaning, it can do about 107,526,041,667 distance calculations per second.

If we have a dataset of 100,000 vectors, and we want to calculate the any-to-any distance, we have 10,000,000,000 calculations to perform, which means that *if written optimally* it should be able to finish at 0.1 seconds.

**Comment by AmitMY** [comment]
And guess what! PyTorch has it built in

```py
# Compute the cosine similarity matrix in a single call
similarities = torch.nn.functional.cosine_similarity(hyp_features,ref_features, dim=-1)
```
or so says chatgpt
https://chatgpt.com/share/67368683-4bb8-800e-8dd6-c37971a1b87c

**Comment by cleong110** [comment]
Nice! 


Here's my own conversation with ChatGPT, which I initialized with the one posted above: https://chatgpt.com/share/67379f5e-2f54-800e-9b76-f55e5d187ed9

[Some quick testing on Colab](https://colab.research.google.com/drive/12DDl6KArHm5hbpvs5mwSAE93aaIKCKpd?usp=sharing) taught me a few things:
* torch is indeed wildly faster. Even if I naively  go pair-by-pair, it's something like 600k/sec instead of 5k/sec. So this can get it done in about an hour or two. Still that's obviously slower than doing a big matrix all in one go.
* So of course, I tried that. But if I populate hyp_features and ref_features with all 3.5 billion combinations I run out of memory and crash.
* So I used a generator to collect it into batches. That massively increases the speed, as in it gets to 55% done in 42 seconds. Of course, if I do it in batches but collect the results in a list... I run out of memory and crash. 
* So I used the cache as suggested. That slows it down significantly again, now taking more like 4-5 minutes. 
* except that I (well, ChatGPT and I) had messed up the keys for the cache-writing so that I couldn't actually get the results back out. I asked it for a version that would write block-by-block and it somehow became slower, now more like 10 minutes. But it apparently worked! Of course if I tried to actually load all this into memory it would crash again
* So I had it yield blocks, instead of trying to concatenate them and return the whole thing. I was successfully able to get the results back out, 8 blocks of shape [10000,83116] and one of [3116,83116].  Not happy about the 10-minute time, but it's way better than 170 hours.

But then it doesn't quite fit the bill for what score_all is supposed to do! It's _yielding_ blocks of scores, rather than _returning_ a list of list of floats as in [the CLIPScore reference](https://github.com/sign-language-processing/signwriting-evaluation/blob/main/signwriting_evaluation/metrics/clip.py#L135-L136C41)

I'm considering how to proceed, and it kind of depends how we want `score_all` to work, and how we intend to run these metrics. Any thoughts?

**Comment by cleong110** [comment]
#5 Closes this.

**Comment by AmitMY** [comment]
I don't think so. #5 is an embedding metric, which is great.
But I expect a SignCLIP metric to extend EmbeddingMetric, and get poses, and run SignCLIP on the poses as part of the pipeline (so this doesn't have to be pre-done). This is for model training for example, where you generate poses on the fly, and now need to test them

**Comment by cleong110** [comment]
Yeah, that would be ideal. Trouble is that running SignCLIP is currently a very manual process involving putting models in the right folders and such. I've been meaning to make a dockerfile or something, but haven't gotten around to it yet. See #12

**Comment by J22Melody** [comment]
Ah, so in #5 we have implemented a base class but it is not really used, right?

We can start with a function that 

1. just returns a dummy number
2. then maybe get the score by the API call, which will not scale unless we maintain the server, but should be enough at this point
3. in the end, can also see how to make an inference code with minimal dependencies, should be possible

**Comment by cleong110** [comment]
Right! In #7 I was planning to make a "load data, run metrics" script, and as part of that adapt [this ASL Citizen one](https://github.com/sign-language-processing/pose-evaluation/blob/main/pose_evaluation/evaluation/evaluate_signclip.py) as a starting point that just loads in .npy and .pose files that are pre-done. 

But I'm not 100% sure whether to load from folders or from `datasets`, so #10. But just to get something started I was thinking to just forge ahead with "folder of .npy files + ASL Citizen split csv"

**Comment by cleong110** [comment]
As for next action on this one, I like the idea of using the API server as a starting point.

**Comment by J22Melody** [comment]
I think it is practical to use pre-computed embeddings for your study at the moment, especially if you have a large dataset and multiple runs to compare.

As for this library, we need to calculate the score given two poses in one run, let's say start with the API. Then we can see how to optimize and scale it.

**Comment by cleong110** [comment]
Implemented with precomputed embeddings for the study, https://github.com/cleong110/pose-evaluation/blob/datasets_to_dataframes/pose_evaluation/metrics/embedding_distance_metric.py
* eval script that used this had to do a bit of a kludge to load the embeddings instead of the .poses: https://github.com/cleong110/pose-evaluation/blob/datasets_to_dataframes/pose_evaluation/evaluation/load_splits_and_run_metrics.py#L215-L221
